vsintegration/src/FSharp.Editor/FSharpBraceMatchingService.fs

```diff
@@ -0,0 +1,114 @@
+        else
+            let completeTextSpan = TextSpan(0, sourceText.Length)
+            let classificationData = FSharpColorizationService.GetColorizationData(sourceText, completeTextSpan, fileName, [], cancellationToken)
+
+            let shouldBeIgnored(characterPosition) =
+                match classificationData.GetClassifiedSpan(characterPosition) with
+                | None -> false
+                | Some(classifiedSpan) -> ignoredClassificationTypes |> Seq.contains classifiedSpan.ClassificationType
+                
+            if shouldBeIgnored(position) then
+                None
+            else
+                let currentCharacter = sourceText.[position]
+
+                let proceedToStartOfString(i) = i - 1
+                let proceedToEndOfString(i) = i + 1
+
+                let afterEndOfString(i) = i >= sourceText.Length
+                let beforeStartOfString(i) = i < 0
+
+                let pickBraceType(leftBrace, rightBrace) =
+                    if currentCharacter = leftBrace then Some(proceedToEndOfString, afterEndOfString, leftBrace, rightBrace)
+                    else if currentCharacter = rightBrace then Some(proceedToStartOfString, beforeStartOfString, rightBrace, leftBrace)
+                    else None
+
+                match supportedBraceTypes |> Seq.tryPick(pickBraceType) with
+                | None -> None
+                | Some(proceedFunc, stoppingCondition, matchedBrace, nonMatchedBrace) ->
+                    let mutable currentPosition = proceedFunc position
+                    let mutable result = None
+                    let mutable braceDepth = 0
+
+                    while result.IsSome = false && stoppingCondition(currentPosition) = false do
+                        cancellationToken.ThrowIfCancellationRequested()
+                        if shouldBeIgnored(currentPosition) = false then
+                            if sourceText.[currentPosition] = matchedBrace then
+                                braceDepth <- braceDepth + 1
+                            else if sourceText.[currentPosition] = nonMatchedBrace then
+                                if braceDepth = 0 then
+                                    result <- Some(BraceMatchingResult(TextSpan(min position currentPosition, 1), TextSpan(max position currentPosition, 1)))
+                                else
+                                    braceDepth <- braceDepth - 1
+                        currentPosition <- proceedFunc currentPosition
+                    result
+
+    interface IBraceMatcher with
+        member this.FindBracesAsync(document: Document, position: int, cancellationToken: CancellationToken): Task<Nullable<BraceMatchingResult>> =
+            let computation() =
+                let sourceText = document.GetTextAsync(cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult()
+                try match getBraceMatchingResult(sourceText, Some(document.Name), position, cancellationToken) with
```

Comment:
Could `FindBracesAsync` use the syntax structure of the `Document` to find matches rather than the raw text? That might allow matching across missing tokens for instance.